### PR TITLE
fix: override clean_html function for anchor tag to save target atrri…

### PR DESCRIPTION
**Pull request by AfaqShuaib09**
[PROD-2089](https://2u-internal.atlassian.net/browse/PROD-3089)

### Description
html2text `handle` method didn't store the anchor tag target attribute. This PR overrides the `handle` method to store the target attribute in the anchor tag.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @AfaqShuaib09!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

File: course_discovery/apps/course_metadata/utils.py [hunk 1]
```diff
@@ -663,6 +665,17 @@ def handle_tag(self, tag, attrs, start):
...
 
+        if tag == 'a':
+            if attrs:
+                attr_dict = dict(attrs)
+                if start and 'href' in attr_dict:
+                    self.outtextf('<a')
+                    for attr, value in attr_dict.items():
+                        self.outtextf(f' {attr}="{value}"')
+                    self.outtextf('>')
+            if not start:
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/utils.py]: Why is this needed if the tag is ignored?

Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/utils.py]: and if HTML cleaning is performed by a library/package, it is possible the package has a way of whitelisting tags and attributes.

Comment (AfaqShuaib09) [reply course_discovery/apps/course_metadata/utils.py]: I overrode the default `handle` function here for the anchor tag, because by default it will just return the `href` and `title` attribute only
so I override it to return all the attributes for the anchor tag
the other workaround, I have found that to override the `handle_a` method, in which I add the logic to include all the tags for `a` (anchor) tag.

Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/utils.py]: Checking options from here, https://github.com/Alir3z4/html2text/blob/master/docs/usage.md, shouldn't ignoring anchor do the job? As a side-effect, we would want to ensure that there are no security repercussions with this. Example https://blog.udemy.com/javascript-href/#:~:text=In%20JavaScript%2C%20you%20can%20call,tag%20(hyperlink%20in%20HTML).

Comment (AfaqShuaib09) [reply course_discovery/apps/course_metadata/utils.py]: If we ignore the anchor tag using ignore_links=True without overriding it, it will simply ignore it and doesn't include anchors tag in output
`input`
image
`output`
image

Comment (AfaqShuaib09) [reply course_discovery/apps/course_metadata/utils.py]: i have tested locally that HTML2Text doesn't perform any security checks in the HTML text.
`input`
```bash
clean_html('<p>please visit this <a title="less go" href="javascript:void(0);" target="_blank" rel="noopener">link</a></p>\n<p>less go</p>'))
```
`output`

```output
<p>please visit this <a href="javascript:void(0);" title="less go">link</a></p>
```
***it didn't remove function call from href tag.***

File: course_discovery/apps/course_metadata/utils.py [hunk 2]
```diff
@@ -646,23 +646,29 @@ class HTML2TextWithLangSpans(html2text.HTML2Text):
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
+        self.ignore_links = True
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/utils.py:649]: How is the data cleaned from the incoming html using this `ignore_links` option?

Comment (AfaqShuaib09) [reply course_discovery/apps/course_metadata/utils.py:649]: we have already removed unecessary tags and spaces using `BeutifulSoup` and `lxml parser` in this `clean_html` to get clean data
https://github.com/openedx/course-discovery/blob/master/course_discovery/apps/course_metadata/utils.py#L667-L677
and by ```self.ignore_links=True``` we simply ignoring call the default handle method call for anchor tag (because by default html2text method doesn't append target attribute to the anchor tag)
https://github.com/Alir3z4/html2text/blob/master/html2text/__init__.py#L525-L538

Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/utils.py:649]: Ohh, right. So this is how our target issue is resolved.. Got it.

**Review by Ali-D-Akbar — APPROVED** [review]

File: course_discovery/apps/course_metadata/utils.py (hunk 1 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/utils.py]: ok, we should be dealing with this too, in a followup PR.

File: course_discovery/apps/course_metadata/utils.py [hunk 3]
```diff
@@ -646,23 +646,29 @@ class HTML2TextWithLangSpans(html2text.HTML2Text):
...
-                    self.outtextf('</span>')
+            if attrs and start and 'lang' in dict(attrs):
+                self.outtextf(f'<span lang="{dict(attrs)["lang"]}">')
+                self.in_lang_span = True
+            if not start and self.in_lang_span:
+                self.outtextf('</span>')
                 self.in_lang_span = False
 
+        if tag == 'a':
+            if attrs and start and 'href' in dict(attrs):
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/utils.py:666]: what does start represent?

File: course_discovery/apps/course_metadata/utils.py [hunk 4]
```diff
@@ -646,23 +646,29 @@ class HTML2TextWithLangSpans(html2text.HTML2Text):
...
-                if self.in_lang_span:
-                    self.outtextf('</span>')
+            if attrs and start and 'lang' in dict(attrs):
+                self.outtextf(f'<span lang="{dict(attrs)["lang"]}">')
+                self.in_lang_span = True
+            if not start and self.in_lang_span:
+                self.outtextf('</span>')
                 self.in_lang_span = False
 
+        if tag == 'a':
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/utils.py:663]: Add a comment to explain why this is needed. Better, make this handling into its own function.

**Review by DawoudSheraz — COMMENTED** [review]
Missing related unit tests

File: course_discovery/apps/course_metadata/utils.py (hunk 3 above)
Comment (AfaqShuaib09) [reply course_discovery/apps/course_metadata/utils.py:666]: `start` is a boolean variable to check whether it is `starting tag` of a `tag` or not

File: course_discovery/apps/course_metadata/utils.py (hunk 4 above)
Comment (AfaqShuaib09) [reply course_discovery/apps/course_metadata/utils.py:663]: Ok i will add the inline comment here. I have followed the same convention here which is followed for span tag above.

**Review by DawoudSheraz — APPROVED** [review]
Squash and rebase before merge

**Comment by openedx-webhooks** [comment]
@AfaqShuaib09 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/course-discovery&entry.867055334=https://github.com/openedx/course-discovery/pull/3752&entry.1484655318=https://github.com/AfaqShuaib09&entry.752974735=2023-01-06+11:24&entry.1917517419=2023-01-11+11:20&entry.2133058324=Yes) so we can improve your experience in the future.
